File: Leetcode/0059 Spiral Matrix II.java
class Solution {
    private static final int[][] dirs = {
            {0, 1},
            {1, 0},
            {0, -1},
            {-1, 0},
    };

    public int[][] generateMatrix(int n) {
        int[][] matrix = new int[n][n];
        int number = 1, elements = n * n, dirIndex = 0;
        int i = 0, j = 0;
        while (number <= elements) {
            matrix[i][j] = number;
            int nextI = i + dirs[dirIndex][0], nextJ = j + dirs[dirIndex][1];
            boolean isInvalid = 
                nextI < 0 || nextI >= n 
                || nextJ < 0 || nextJ >= n 
                || (matrix[nextI][nextJ] != 0 && matrix[nextI][nextJ] < number);
            if (isInvalid) {
                dirIndex = (dirIndex + 1) % dirs.length;
            }
            i += dirs[dirIndex][0];
            j += dirs[dirIndex][1];
            ++number;
        }
        return matrix;
    }
}

File: Leetcode/0059 Spiral Matrix II.ts
const DIRS = [[0, 1], [1, 0], [0, -1], [-1, 0]];
const EMPTY = -1;

function generateMatrix(n: number): number[][] {
    const matrix = new Array<number[]>(n);
    for (let i = 0; i < n; i++) {
        matrix[i] = new Array<number>(n).fill(EMPTY);
    }
    const nSquare = n * n;
    let i = 0, j = 0, dirIndex = 0, num = 1;
    while (num <= nSquare) {
        matrix[i][j] = num++;
        const nextI = i + DIRS[dirIndex][0], nextJ = j + DIRS[dirIndex][1];
        const changeDir = nextI < 0 || nextI >= n || nextJ < 0 || nextJ >= n || matrix[nextI][nextJ] !== EMPTY;
        if (changeDir) {
            dirIndex = (dirIndex + 1) % DIRS.length;
        }
        i += DIRS[dirIndex][0];
        j += DIRS[dirIndex][1];
    }
    return matrix;
}
